### PR TITLE
Add controllers and OCM add-ons that will enable Argo Workflow multicluster capability

### PR DESCRIPTION
**Pull request by mikeshng**
Add Placement controller to evaluate Workflow Placement. Add ManifestWork controller to propagate Workflow to ManagedClusters. Add Status controller to sync remote cluster Workflow status back to hub cluster. Add Install add-on to automate Argo Workflow deployments to all managed clusters. Add Workflow status sync add-on to sync Worfklow status.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,122 @@
...
+the [Open Cluster Management (OCM)](https://open-cluster-management.io/) APIs and components.
+
+## Description
+By using this project, users can propagate Argo Workflow to remote clusters based on availabile cluster resource usages.
+
+![multi-cluster](assets/multicluster.png)
+
+There are multiple components in this project.
+
+- Placement controller watches the Workflow CR and evaluate the [Placement](https://open-cluster-management.io/concepts/placement/) decision then annotates the Workflow with the remote cluster target.
```
Comment (juliev0) [inline README.md]: 1. "evaluate" -> "evaluates"
2. add comma before "then"

File: README.md [hunk 2]
```diff
@@ -1 +1,122 @@
...
+- ManifestWork controller watches the Workflow CR with the remote cluster target and create [ManifestWork](https://open-cluster-management.io/concepts/manifestwork/) CR which will propagate the Workflow to the intended remote cluster.
+- Status controller that updates the Workflow CR status on the hub cluster with the remote cluster's Workflow's execution results.
+- Install Add-on that automates the installation of Argo Workflow to all the managed clusters.
+See the [Install Add-on README](addons/hub/install/README.md) for more details.
+- Workflow Status Sync Add-on that sync the entire Argo Workflow status from managed clusters to hub cluster
+See the [Status Sync Add-on README](addons/hub/status_sync/README.md) for more details.
+
+See the [Workflow example](example/hello-world.yaml) for the required label(s) and annotation(s).
```
Comment (juliev0) [inline README.md:20]: 1. line 13: "create" -> "creates"
2. line 15: "Argo Workflow" -> "Argo Workflows"
3. line 17: "sync" -> "syncs"
4. line 23: both instances of "setup" -> "set up"

File: README.md [hunk 3]
```diff
@@ -1 +1,122 @@
...
+    type: Completed
+  finishedAt: "2022-11-12T21:57:19Z"
+...
+  phase: Succeeded
+  progress: 1/1
+  resourcesDuration:
+    cpu: 9
+    memory: 9
+...
+```
```
Comment (juliev0) [inline README.md:106]: 1. Line 40: "Setup" -> "Set up"
2. Line 40: "registered" -> "register"
3. Line 51 and 52: "Argo Workflows" -> "Argo Workflows"

File: addons/hub/install/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# Argo Workflow Installation Add-on
+By applying this add-on to your OCM hub cluster, the Argo Workflow installation will automatically be applied to all your existing `managed` (`spoke`) clusters and all your to be registered `managed` (`spoke`) clusters.
+
+Under the `manifests` folder it contains the Argo Workflow install version of:
+
+```
+v3.4.2
+```
+
```
Comment (juliev0) [inline addons/hub/install/README.md:9]: Line 2 and 4: "Argo Workflow" -> "Argo Workflows"
Line 2: "to be registered" -> "to-be-registered"
Line 12: "Setup" -> "Set up"

File: addons/hub/status_sync/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+argoworkflow-status-sync-addon-agent    1/1     1            1           2m24s
+```
+
+You can also validate and check the status of the add-on on the `Hub` cluster:
+
+```
+$ kubectl -n cluster1 get managedclusteraddon # replace "cluster1" with your managed cluster name
+NAME                                AVAILABLE   DEGRADED   PROGRESSING
+argoworkflow-status-sync-addon      True                   
+```
```
Comment (juliev0) [inline addons/hub/status_sync/README.md:40]: 1. Line 2: "be sync" -> "be synced"
2. Line 22: "Setup" -> "Set up"

File: controllers/workflow/workflow_controller.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	argov1alpha1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	workflowv1alpha1 "github.com/sarabala1979/argo-wf-ocm-multicluster/api/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	workv1 "open-cluster-management.io/api/work/v1"
+)
+
+const (
+	// Workflow annotation that dictates which managed cluster this Workflow should be propgated to.
```
Comment (juliev0) [inline controllers/workflow/workflow_controller.go]: "propagated"

File: controllers/workflow/workflow_placement_controller.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+// SetupWithManager sets up the controller with the Manager.
+func (re *WorkflowPlacementReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&argov1alpha1.Workflow{}).
+		WithEventFilter(WorkflowPlacementPredicateFunctions).
+		Complete(re)
+}
+
+// Reconcile evaluates the PlacementDecision based on the Placement reference then populates the ManagedCluster annotation with the reuslt
```
Comment (juliev0) [inline controllers/workflow/workflow_placement_controller.go]: "result"

File: controllers/workflow/workflow_status_controller.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		For(&workflowv1alpha1.WorkflowStatusResult{}).
+		Complete(re)
+}
+
+// Reconcile populates the Workflow status based on the associated WorkflowStatusResult
+// The status sync flow:
+// Workflow (dormant) on hub cluster is created and it will be propagated to managed cluster(s)
+// => Workflow on managed cluster (contains annotations that reference the hub cluster dormant Workflow)
+// => The managed cluster status sync agent will create/update a WorkflowStatusResult on the hub cluster (contains annotations that reference the hub cluster dormant Workflow)
+// => using the references from WorkflowStatusResult this reconciler finds the dormant Workflow and populate the status.
```
Comment (juliev0) [inline controllers/workflow/workflow_status_controller.go]: "populate" -> "populates"

**Review by juliev0 — APPROVED** [review]

File: config/crds/workflows_crd.yaml [hunk 9]
```diff
@@ -0,0 +1,55 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (sarabala1979) [inline config/crds/workflows_crd.yaml]: Can we refer to the release version from the argo workflow repo?

Comment (mikeshng) [reply config/crds/workflows_crd.yaml]: Done as suggested. Thanks.

**Comment by sarabala1979** [comment]
@mikeshng I will review it today

**Review by sarabala1979 — APPROVED** [review]

**Pull request merged**